assignment5/AlphabetDiscovering.java

```diff
@@ -0,0 +1,107 @@
+                for (int i = 1; i < words.size(); ++i) {
+                    String firstWord = words.get(i - 1);
+                    String secondWord = words.get(i);
+                    int minLength = Math.min(firstWord.length(), secondWord.length());
+                    for (int j = 0; j < minLength; ++j) {
+                        if (firstWord.charAt(j) != secondWord.charAt(j)) {
+                            ordering.onOrderedPair(firstWord.charAt(j), secondWord.charAt(j));
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /*
+     * Class to solve the problem, which does topological sort in existing graph
+     */
+    public static class Ordering {
+        private final Map<Character, Vertex> vertexMap = new HashMap<>();
+
+        /*
+         * Method to start the process of ordering. It scans the given wordlist and creates a graph
+         * --------------------------------------------------------
+         * @param words - given dictionary in unknown language
+         * --------------------------------------------------------
+         * @return initialised Ordering object with a graph
+         */
+        public static Ordering seedWords(List<String> words) {
+            Ordering o = new Ordering();
+            LetterScanner letterScanner = new LetterScanner();
+            letterScanner.scan(words, o);
+            return o;
+        }
+
+        /*
+         * Method to do topological sort of the graph
+         * ------------------------------------------
+         * @return list of the found characters - unknown alphabet
+         */
+        public List<Character> topologicalOrdering() {
+            Set<Vertex> unmarked = new HashSet<>(vertexMap.values());
+            Set<Vertex> visiting = new HashSet<>();
+            Set<Vertex> visited = new HashSet<>();
+
+            List<Character> ordered = new LinkedList<>();
+            while (!unmarked.isEmpty()) {
+                Vertex v = unmarked.iterator().next();
+                visitVertex(v, unmarked, visiting, visited, ordered);
+            }
+
+            Collections.reverse(ordered);
+            return ordered;
+        }
+
+        /*
+         * Method, which does the main part of topological sort.
+         * It visits vertexes of the given graph.
+         * ------------------------------------------------
+         * @param v - vertex to visit
+         * @param unmarked - set of unmarked vertex
+         * @param visiting - set of vising vertex
+         * @param visited - set of visited vertex
+         * @ordered - list of ordered characters
+         */
+        private static void visitVertex(Vertex v, Set<Vertex> unmarked, Set<Vertex> visiting,
+                                  Set<Vertex> visited, List<Character> ordered) {
+            if (visiting.contains(v)) {
+                throw new RuntimeException("Cyclic graph. No ordering possible.");

```

Comment:
InvalidStateException or even IllegalArgumentException would be better here - it would provide clear information about the reason of failure rather than generic "something's wrong at runtime".